torch/csrc/distributed/rpc/rref.h

```diff
@@ -13,71 +14,236 @@ namespace rpc {
+//
+// Note that the notification might come delayed or out-of-order.
+//
+// With #2 and #3, it is possible that the owner only partially knows the RRef
+// fork graph or not even knowing it at all. For example, the RRef could be
+// constructed on a user, and before the owner receives the RPC call, the
+// creator user might have already shared the RRef with other users, and those
+// users could further share the RRef. One invariant is that the fork graph of
+// any RRef is always a tree rooted at the owner, because forking an RRef always
+// creates a new RRef instance, and hence every RRef has a single parent. One
+// nasty detail is that when an RRef is created on a user, technically the owner
+// is not its parent but we still consider it that way and it does not break the
+// argument below.
+//
+// The owner's view on any node (fork) in the tree has three stages:
+//            1) unknown → 2) known → 3) deleted.
+// The owner's view on the entire tree keeps changing. The owner deletes its
+// OwnerRRef instance when it thinks there are no living UserRRefs, i.e., all
+// the UserRRefs could be either indeed deleted or unknown. The dangerous case
+// is when some forks are unknown and others are deleted. We only need a simple
+// guarantee to prevent this situation:
+//
+// * G2. Parent UserRRef cannot be deleted until the child UserRRef is confrimed
+//       by the owner.
+//
+// G2 trivially guarantees that no parent UserRRef Y can be deleted before the
+// owner knows all Y's children UserRRefs.
+//
+// However, it is possible that a child UserRRef Z is deleted before the owner
+// knows its Z's parent Y. More specifically, this can happen when Z's
+// RREF_FORK_REQUEST was processed by the owner before any other messages from
+// Y, where Z receives RREF_USER_ACCEPT from the owner and then send out
+// RREF_USER_DELETE before the owner leanrs about Y. Nevertheless, this does not
+// cause any problem. Because, at least one of Y's ancestor will be alive if Z,
+// preventing the owner from deleting the OwnerRRef. Consider the following
+// example:
+//
+//    OwnerRRef -> A -> Y -> Z
+//
+// OwnerRRef forks to A, then A forks to B, B forks to Y, and Y forks to Z. Z
+// can be deleted without OwnerRRef knowing Y or even B. However, the OwnerRRef
+// will at least know A, A won't die before the owner knows Y.
+//
+// Things get a little trickier if the RRef starts from a user:
+//
+// OwnerRRef
+//     ^
+//     |
+//     A -> Y -> Z
+//
+// If Z calls to_here on the UserRRef, the owner at least knows A when Z is
+// deleted, because otherwise to_here wouldn't finish. If Z does not call
+// to_here, it is possible that the owner receives all messages from Z before
+// any message from A and Y. In this case, as the OwnerRRef hasn't been created
+// on the owner yet (it is only created when owner receives a remote call or a
+// to_here() call), there is nothing to be deleted either. Hence, it's still OK.
+//
+//
```

Comment:
Didn't review the entire comment. I assume it's the same as the issue you posted yesterday with a full description of the RRef design?

Reply:
This is a little obsolete, let me update it with the latest version.